Project/SubZeroDriver/SubZero.cpp

```diff
@@ -0,0 +1,456 @@
+			KdPrint((DRIVER_PREFIX "[-] Failed to create sym link (status=0x%08X)\n",status));
+			break;
+		}
+		symLinkCreated = true;
+
+		// Register for object notifications
+		status = ::ObRegisterCallbacks(&reg, &g_Globals.ObjectRegistrationHandle);
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "[-] Failed to register object callbacks (status=0x%08X)\n", status));
+			break;
+		}
+
+		status = ::CmRegisterCallbackEx(OnRegistryNotify, &altitude, DriverObject,
+			nullptr, &g_Globals.RegistryRegistrationCookie, nullptr);
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "failed to set registry callback (%08X)\n",
+				status));
+			break;
+		}
+
+		// Register for Thread notifications
+		status = ::PsSetCreateThreadNotifyRoutine(OnThreadNotify);
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "[-] Failed to set Thread callbacks (status=status=%08X)\n", status));
+			break;
+		}
+	} while (false);
+	
+	if (!NT_SUCCESS(status)) {
+		if (symLinkCreated)
+			::IoDeleteSymbolicLink(&symLink);
+		if (DeviceObject)
+			::IoDeleteDevice(DeviceObject);
+	}
+
+	// Initialize RundownProtection for APCs
+	::ExInitializeRundownProtection(&g_Globals.RundownProtection);
+
+	// Search for explorer Process
+	PEPROCESS explorerProcess;
+	LARGE_INTEGER interval;
+	interval.QuadPart = -50000000; // 5 Seconds / 100 nanoseconds - in RELATIVE time
+	do {
+		if (::FindProcessByName(PARENT_PROCESS_NAME, &explorerProcess)) {
+			g_Globals.ExplorerPID = ::HandleToULong(::PsGetProcessId(explorerProcess));
+			KdPrint((DRIVER_PREFIX "[+] explorer.exe found. PID: %d\n", g_Globals.ExplorerPID));
+			//ChangeVadEntryProtection(explorerProcess, NULL, 0);
+			break;
+		}
+		else { 
+			KdPrint((DRIVER_PREFIX "[-] explorer.exe not found. Trying again in 5 seconds\n"));
+			::KeDelayExecutionThread(KernelMode, false, &interval); }
+	} while (true);
+
+	//::InterlockedExchange((volatile LONG*)((uintptr_t)explorerProcess + 0x3e8) , 1);
+	//KdPrint((DRIVER_PREFIX "[+] explorer.exe PID changed.")); 
+
+	DriverObject->DriverUnload = SubZeroUnload;
+	DriverObject->MajorFunction[IRP_MJ_CREATE] = DriverObject->MajorFunction[IRP_MJ_CLOSE] = SubZeroCreateClose;
+	return status;
+}
+
+void SubZeroUnload(PDRIVER_OBJECT DriverObject) {
+	// unregister Process & Thread notifications in case of an error
+	::PsSetCreateProcessNotifyRoutineEx(OnProcessNotify, TRUE);
+	::PsRemoveCreateThreadNotifyRoutine(OnThreadNotify);
+
+	// unregister object & registry notifications
+	::ObUnRegisterCallbacks(g_Globals.ObjectRegistrationHandle);
+	::CmUnRegisterCallback(g_Globals.RegistryRegistrationCookie);
+
+	// Wait for KAPC to finish in case of an error
+	::ExWaitForRundownProtectionRelease(&g_Globals.RundownProtection);
+
+	// Delete device
+	UNICODE_STRING symLink = RTL_CONSTANT_STRING(L"\\??\\" DRIVER_NAME);
+	::IoDeleteSymbolicLink(&symLink);
+	::IoDeleteDevice(DriverObject->DeviceObject);
+
+	KdPrint((DRIVER_PREFIX "[+] Driver unloaded successfully\n"));
+}
```

Comment:
You could (in theory) create RAII wrapper class for registering notifications.
I didn't done it but it could be good practice and challenge.
Maybe we will design it together.